### PR TITLE
Add Apache Pulsar e2e encryption

### PR DESCRIPTION
**Pull request by yaron2**
Closes https://github.com/dapr/components-contrib/issues/2531.

### CODE REVIEW

**Review by DeepanshuA — APPROVED** [review]
LGTM. I think we should attach a docs issue with this one, to let users know the possibility of encryption/decryption with Pulsar.

File: pubsub/pulsar/pulsar.go [hunk 1]
```diff
@@ -217,6 +226,14 @@ func (p *Pulsar) Publish(ctx context.Context, req *pubsub.PublishRequest) error
 			opts.Schema = getPulsarSchema(sm)
 		}
 
+		if p.useProducerEncryption() {
+			keyReader := crypto.NewFileKeyReader(p.metadata.PublicKey, "")
```
Comment (ItalyPaleAle) [inline pubsub/pulsar/pulsar.go:230]: Please also provide a way for the keys themselves to be passed as value for the metadata properties, not just a path to a file to disk (eg the value of the property being the actual PEM-encoded string). We do the same for other components such as the HTTP binding.

**Review by ItalyPaleAle — DISMISSED** [review]

File: pubsub/pulsar/pulsar.go (hunk 1 above)
Comment (yaron2) [reply pubsub/pulsar/pulsar.go:230]: This isn't possible as the Pulsar SDK only takes file paths.

Comment (berndverst) [reply pubsub/pulsar/pulsar.go]: I agree with that concern - mounting files into containers in various container environments is really a hassle that is best avoided.

**Comment by berndverst** [comment]
PR seems fine now but certification test is failing. I just restarted it.

**Comment by yaron2** [comment]
Passing now @berndverst

**Comment by DeepanshuA** [comment]
Cert tests seem to be failing: === FAIL: pubsub/pulsar TestPulsarEncryptionFromData/pulsar_encryption_test_with_data (unknown)

ref:
https://github.com/dapr/components-contrib/actions/runs/4829205810/jobs/8603965674?pr=2812
https://github.com/dapr/components-contrib/actions/runs/4825283856/jobs/8595995151?pr=2812

**Comment by yaron2** [comment]
> Cert tests seem to be failing: === FAIL: pubsub/pulsar TestPulsarEncryptionFromData/pulsar_encryption_test_with_data (unknown)
> 
> ref: https://github.com/dapr/components-contrib/actions/runs/4829205810/jobs/8603965674?pr=2812 https://github.com/dapr/components-contrib/actions/runs/4825283856/jobs/8595995151?pr=2812

Was transient. All tests are passing. cc @berndverst

**Review by berndverst — APPROVED** [review]

**Comment by hugosmitter** [comment]
This is fantastic. Thank you so much to the Dapr community for adding this feature to the Dapr PubSub component for Pulsar.

**Pull request merged**